core/src/main/java/com/sequenceiq/distrox/v1/distrox/StackOperations.java

```diff
@@ -237,22 +237,21 @@ public FlowIdentifier updatePillarConfiguration(@NotNull NameOrCrn nameOrCrn, Lo
     }
 
     public UpgradeV4Response checkForClusterUpgrade(@NotNull NameOrCrn nameOrCrn, Long workspaceId, UpgradeV4Request request) {
-        if (nameOrCrn.hasName()) {
-            String stackName = nameOrCrn.getName();
-            boolean osUpgrade = upgradeService.isOsUpgrade(request);
-            UpgradeV4Response upgradeResponse = clusterUpgradeAvailabilityService.checkForUpgradesByName(workspaceId, stackName,
-                    osUpgrade);
-            if (CollectionUtils.isNotEmpty(upgradeResponse.getUpgradeCandidates())) {
-                clusterUpgradeAvailabilityService.filterUpgradeOptions(upgradeResponse, request);
-            }
-            Stack stack = getStackByName(stackName);
-            MDCBuilder.buildMdcContext(stack);
+        Stack stack = stackService.getByNameOrCrnInWorkspace(nameOrCrn, workspaceId);
+        MDCBuilder.buildMdcContext(stack);
+        boolean osUpgrade = upgradeService.isOsUpgrade(request);
+        UpgradeV4Response upgradeResponse = clusterUpgradeAvailabilityService.checkForUpgradesByName(stack, osUpgrade, request.getReplaceVms());
+        if (CollectionUtils.isNotEmpty(upgradeResponse.getUpgradeCandidates())) {
+            clusterUpgradeAvailabilityService.filterUpgradeOptions(upgradeResponse, request);
+        }
+        validateDatalakeHasNoDatahub(workspaceId, stack, upgradeResponse);
+        return upgradeResponse;
+    }
+
+    private void validateDatalakeHasNoDatahub(Long workspaceId, Stack stack, UpgradeV4Response upgradeResponse) {
```

Comment:
We're checking if the attached Data Hubs are stopped or not so the method name is misleading. Similar we should rename `checkForNotAttachedClusters` as well.

Reply:
renamed